vmdb/spec/automation/unit/engine_validation/miq_ae_state_machine_retry_spec.rb

```diff
@@ -70,36 +90,51 @@ def create_state_class(attrs = {})
                                    'ae_instances' => ae_instances))
   end
 
-  it "check retry" do
-    setup_model
-    MiqServer.stub(:my_zone).and_return('default')
-    @expected = {'three' => 3, 'one'  => 1, 'two'  => 2, 'gravy' => 'train'}
-    args = {
-      :namespace        => @namespace,
-      :class_name       => @state_class,
-      :instance_name    => @state_instance,
-      :automate_message => 'create'
-    }
-    q = MiqQueue.put(
-                    :role        => 'automate',
-                    :class_name  => 'MiqAeEngine',
-                    :method_name => 'deliver',
-                    :args        => [args])
-    delivered_ids = []
-    status, message, ws = q.deliver
+  def create_root_class(attrs = {})
+    ae_fields = {'rel1' => {:aetype => 'relationship', :datatype => 'string'}}
+    fqname = "/#{@domain}/#{@namespace}/#{@state_class}/#{@state_instance}"
+    ae_instances = {@root_instance => {'rel1' => {:value => fqname}}}
+    FactoryGirl.create(:miq_ae_class, :with_instances_and_methods,
+                       attrs.merge('ae_fields'    => ae_fields,
+                                   'ae_methods'   => {},
+                                   'ae_instances' => ae_instances))
+  end
+
+  it "check persistent hash" do
+    setup_model(method_script_state_var)
+    expected = {'three' => 3, 'one'  => 1, 'two'  => 2, 'gravy' => 'train'}
+    send_ae_request_via_queue(@automate_args)
+    status, _message, ws = deliver_ae_request_from_queue
     status.should_not eq(MiqQueue::STATUS_ERROR)
     ws.should_not be_nil
-    delivered_ids << q.id
     MiqQueue.count.should eq(2)
-    MiqQueue.all.each do |v|
-      unless delivered_ids.include?(v.id)
-        status, message, ws = v.deliver
-        status.should_not eq(MiqQueue::STATUS_ERROR)
-        ws.should_not be_nil
-        delivered_ids << v.id
-        ws.persist_state_hash.should eq(@expected)
-        ws.root.attributes['finished'].should be_true
-      end
+    status, _message, ws = deliver_ae_request_from_queue
+    status.should_not eq(MiqQueue::STATUS_ERROR)
+    ws.persist_state_hash.should eq(expected)
+    ws.root.attributes['finished'].should be_true
+  end
+
+  it "check max retries" do
+    setup_model(perpetual_retry_script)
+    send_ae_request_via_queue(@automate_args)
+    (@max_retries + 2).times do
+      status, _message, ws = deliver_ae_request_from_queue
+      status.should_not eq(MiqQueue::STATUS_ERROR)
+      ws.should_not be_nil
+    end
+    deliver_ae_request_from_queue.should be_nil
+  end
+
+  it "check max_time" do
+    setup_model(perpetual_retry_script)
+    send_ae_request_via_queue(@automate_args)
+    2.times do
+      status, _message, ws = deliver_ae_request_from_queue
+      status.should_not eq(MiqQueue::STATUS_ERROR)
+      ws.should_not be_nil
+      sleep(@max_time)
+      sleep(1)
```

Comment:
Need to investigate using timecop here.  Adding 6 seconds of sleep time will not go over well with @jrafanie.